features/mbedtls/targets/TARGET_STM/sha256_alt.c

```diff
@@ -0,0 +1,141 @@
+    volatile unsigned char *p = v; while( n-- ) *p++ = 0;
+}
+
+void mbedtls_sha256_init( mbedtls_sha256_context *ctx )
+{
+    mbedtls_zeroize( ctx, sizeof( mbedtls_sha256_context ) );
+
+    /* Enable HASH clock */
+    __HAL_RCC_HASH_CLK_ENABLE();
+}
+
+void mbedtls_sha256_free( mbedtls_sha256_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+    /* Force the HASH Periheral Clock Reset */
+    __HAL_RCC_HASH_FORCE_RESET();
+
+    /* Release the HASH Periheral Clock Reset */
+    __HAL_RCC_HASH_RELEASE_RESET();
+
+    mbedtls_zeroize( ctx, sizeof( mbedtls_sha256_context ) );
+}
+
+void mbedtls_sha256_clone( mbedtls_sha256_context *dst,
+                         const mbedtls_sha256_context *src )
+{
+    *dst = *src;
+}
+
+void mbedtls_sha256_starts( mbedtls_sha256_context *ctx, int is224 )
+{
+    /* HASH IP initialization */
+    if (HAL_HASH_DeInit(&ctx->hhash_sha256) == HAL_ERROR) {
+        // error found to be returned
+        return;
+    }
+
+    ctx->is224 = is224;
+    /* HASH Configuration */
+    ctx->hhash_sha256.Init.DataType = HASH_DATATYPE_8B;
+    if (HAL_HASH_Init(&ctx->hhash_sha256) == HAL_ERROR) {
+        // error found to be returned
+        return;
+    }
+}
+
+void mbedtls_sha256_process( mbedtls_sha256_context *ctx, const unsigned char data[MBEDTLS_SHA256_BLOCK_SIZE] )
+{
+    if (ctx->is224 == 0) {
+        HAL_HASHEx_SHA256_Accumulate(&ctx->hhash_sha256, (uint8_t *) data, MBEDTLS_SHA256_BLOCK_SIZE);
+    } else {
+        HAL_HASHEx_SHA224_Accumulate(&ctx->hhash_sha256, (uint8_t *) data, MBEDTLS_SHA256_BLOCK_SIZE);
+    }
+}
+
+void mbedtls_sha256_update( mbedtls_sha256_context *ctx, const unsigned char *input, size_t ilen )
+{
+    size_t currentlen = ilen;
+    // store mechanism to handle MBEDTLS_SHA256_BLOCK_SIZE bytes per MBEDTLS_SHA256_BLOCK_SIZE bytes
+    if (currentlen == 0){ // only change HW status is size if 0
+        if(ctx->hhash_sha256.Phase == HAL_HASH_PHASE_READY) {
+            /* Select the SHA256 or SHA224 mode and reset the HASH processor core, so that the HASH will be ready to compute
+             the message digest of a new message */
+            if (ctx->is224 == 0) {
+                HASH->CR |= HASH_ALGOSELECTION_SHA256 | HASH_CR_INIT;
+            } else {
+                HASH->CR |= HASH_ALGOSELECTION_SHA224 | HASH_CR_INIT;
+            }
+        }
+        ctx->hhash_sha256.Phase = HAL_HASH_PHASE_PROCESS;
+    } else if (currentlen < (MBEDTLS_SHA256_BLOCK_SIZE - ctx->sbuf_len)) {
+        // only buffurize
+        memcpy(ctx->sbuf + ctx->sbuf_len, input, currentlen);
+        ctx->sbuf_len += currentlen;
+    } else {
+        // fill buffer and process it
+        memcpy(ctx->sbuf + ctx->sbuf_len, input, (MBEDTLS_SHA256_BLOCK_SIZE-ctx->sbuf_len));
+        currentlen -= (MBEDTLS_SHA256_BLOCK_SIZE - ctx->sbuf_len);
+        mbedtls_sha256_process(ctx, ctx->sbuf);
+        // now process every input as long as it is %4 bytes
+        size_t iter = currentlen / 4;
```

Comment:
I am slightly confused why this is using the value `4` instead of `64` (or `MBEDTLS_SHA256_BLOCK_SIZE`). As I understand from the original mbed TLS implementation, blocks are processed in 64-byte blocks. Similarly, the documentation for `HAL_HASHEx_SHA256_Accumulate()` states:

>If the Size is not multiple of 64 bytes, the padding is managed by hardware.

I believe that the problem is that if at the start of the function `sbuf_len` is, say `0`, and `ilen` is `69` then the first block will be as expected, but then the second block of 5 bytes will be padded by the hardware to reach 64-bytes. Then if the user calls `mbedtls_sha256_update()` this will yield something like `SHA256(64 || 5 || padding || other bytes)` instead of `SHA256(64 || 5 || otherbytes)`. Perhaps I am missing something though...